app/models/concerns/browseable.rb

```diff
@@ -0,0 +1,157 @@
+module Browseable
+  extend ActiveSupport::Concern
+
+  included do
+    class_attribute :facets
+    self.facets = {}
+  end
+
+  class Facets
+    include Enumerable
+
+    attr_reader :search, :klass
+
+    def initialize(search)
+      @search, @klass = search, search.klass
+    end
+
+    def [](key)
+      facets[key]
+    end
+
+    def key?(key)
+      facet?(key)
+    end
+    alias_method :has_key?, :key?
+
+    def keys
+      klass.facets.keys
+    end
+
+    def each(&block)
+      keys.each do |key|
+        yield key, self[key]
+      end
+    end
+
+    private
+
+    def facet(key)
+      klass.facets.fetch(key)
+    end
+
+    def facet?(key)
+      klass.facets.key?(key)
+    end
+
+    def facets
+      @facets ||= Hash.new(&facet_query)
+    end
+
+    def facet_query
+      lambda do |hash, key|
+        unless facet?(key)
+          raise ArgumentError, "Unsupported facet: #{key.inspect}"
+        end
+
+        hash[key] = scope(key).count
+      end
+    end
+
+    def scope(key)
+      search.klass.instance_exec(&facet(key))
+    end
+  end
```

Comment:
There's a lot of things called ``/facets?/`` in here.  Some that belong to this class, but some that belong to the ``klass``. I wonder if we could rename or delegate to make it clearer what each ``facet`` is.  How does this strike you?

```ruby
class Facets
  include Enumerable

  attr_reader :klass

  def initialize(klass)
    @klass = @klass
  end

  def [](key)
    facet_counts[key]
  end

  delegate :key?, :has_key?, :keys, to: :facet_definitions

  def each(&block)
    keys.each do |key|
      yield key, self[key]
    end
  end

  private
  
  def facet_definitions
    klass.facets
  end

  def facet_counts
    @facets ||= Hash.new(&count_facet_query)
  end

  def count_facet_query
    lambda do |hash, key|
      unless key?(key)
        raise ArgumentError, "Unsupported facet: #{key.inspect}"
      end

      hash[key] = facet_scope(key).count
    end
  end

  def facet_scope(key)
    klass.instance_exec(&facet_definitions[key])
  end
end
```

Not that I'm super happy that all these methods are now ``facet_`` prefixed - it's probably redundant, but leaving them as ``definitions`` or ``counts`` doesn't seem enough...